co-log/src/Colog/Concurrent.hs

```diff
@@ -0,0 +1,129 @@
+       , runInBackgroundThread
+       -- $background-worker
+       , BackgroundWorker
+       , forkBackgroundLogger
+       , killBackgroundLogger
+       , convertToLogAction
+       ) where
+
+import Control.Concurrent
+import Control.Concurrent.STM.TBQueue
+
+import Colog.Core.Action (LogAction (..))
+import Colog.Concurrent.Internal
+
+-- $worker-thread
+-- Worker thread approach allows to create a single worker
+-- thread that will perform all logging. This approach allows
+-- user to use different loggers and change them on the flight.
+--
+-- With bracket resource control style code may look as
+-- following:
+--
+-- @
+-- bracket ('mkBackgroundThread' 4096)
+--         ('killBackgroundThread')
+--         $ \thread -> 
+--            let la1 = runInBackgroundThread worker logger1
+--                la2 = runInBackgroundThread worker logger1
+--            unLogger la1 msg1
+--            unLogger la2 msg2
+--            ..
+-- @
+--
+-- This code will run a background thread and process different
+-- loggers in that thread.
+--
+-- You may extend the logger and entire logger will be run
+-- on the background thread. It is not a case with
+-- 'forkBackgroundLogger'.
+
+-- | Create a background worker with a given capacity.
+-- If capacity is reached, then the thread that tries to
+-- write logs will be blocked.
+mkBackgroundThread :: Int -> IO (BackgroundWorker (IO ()))
+mkBackgroundThread capacity = do
+  queue <- newTBQueueIO (fromIntegral capacity)
+  tid <- forkIO $ forever $ join $ atomically $ readTBQueue queue
+  pure $ BackgroundWorker tid (writeTBQueue queue)
+
+-- | Run logger action asynchronously in the worker thread.
+-- Logger is executed in the other thread entirely, so if
+-- logger takes any thread related context it will be
+-- read from the other thread.
+runInBackgroundThread :: BackgroundWorker (IO ()) -> LogAction IO msg -> LogAction IO msg
+runInBackgroundThread bt logAction = LogAction $ \msg -> do
+  atomically $ backgroundWorkerWrite bt $ unLogAction logAction msg
+
+-- $background-logger
+--
+-- Background logger is specialized version of the Worker 
+-- process. Instead of running any user's logger it will
+-- accept @msg@ type instead and process it with a single
+-- logger defined at creation time. This approach may be more
+-- efficient comparing to the 'runInBackgroundThread' but it's less
+-- generic.
+--
+-- @
+-- bracket ('forkBackgroundLogger' 4096 logger1)
+--         ('killBackgroundThread')
+--         $ \logger-> 
+--            let la = convertToLogAction logger
+--            unLogger la msg1
+--            ..
+-- @
+--
+-- You may extend the logger and but all extensions will be run
+-- in the user thread, only @logger1@ is run in background.
+-- It is not a case with 'runInBackgroundThread'.
+
+-- | Creates background logger with given @Capacity@,
+-- takes a 'LogAction' that should describe how to write
+-- logs.
+--
+-- @capability@ - parameter tells how many in flight messages
```

Comment:
This variable called `capacity` in multiple places, so probably it's a good idea to name it `capacity` here as well instead of `capability`.

Reply:
thanks!